### PR TITLE
[consensus] Add universal committer tests

### PR DESCRIPTION
**Pull request by arun-koshy**
Most test scenarios are identical to the test in base_committer_tests.rs but just going through the `UniversalCommitter` logic to handle checking between direct & indirect rules. Feel free to review all tests but the new tests that were not relevant in base_committer_tests.rs are
- `direct_commit_late_call`
- `no_genesis_commit`
- `no_leader`

Will follow up in another PR with the tests for multi leader and pipelining.

### CODE REVIEW

**Review by akichidis — APPROVED** [review]

File: consensus/core/src/tests/universal_committer_tests.rs [hunk 1]
```diff
@@ -55,6 +56,272 @@ fn direct_commit() {
...
+        None,
+        decision_round_wave_2,
+    );
+
+    // Commit one block.
+    let last_decided = Slot::new_for_test(0, 0);
+    let first_sequence = committer.try_commit(last_decided);
+    assert_eq!(first_sequence.len(), 1);
+
+    // Ensure we don't commit it again.
```
[the remark this answers is not in the document]
Comment (arun-koshy) [reply consensus/core/src/tests/universal_committer_tests.rs:86]: Just to confirm the expected behavior you are looking for. Below we do call `try_commit` with the leader that was decided in the first sequence to ensure that its not committed again. However if you are trying to ensure that calling try commit with the same input results in the same output then I will add that case in here as well.

File: consensus/core/src/tests/universal_committer_tests.rs [hunk 2]
```diff
@@ -57,6 +58,272 @@ fn direct_commit() {
...
+        context.clone(),
+        dag_state.clone(),
+        None,
+        decision_round_wave_2,
+    );
+
+    // Commit one block.
+    let last_decided = Slot::new_for_test(0, 0);
+    let first_sequence = committer.try_commit(last_decided);
+    assert_eq!(first_sequence.len(), 1);
```
Comment (mwtian) [inline consensus/core/src/tests/universal_committer_tests.rs:86]: Add a TODO to verify the round and authority index of committed leader? And which leader is committed here?

File: consensus/core/src/tests/universal_committer_tests.rs [hunk 3]
```diff
@@ -57,6 +58,272 @@ fn direct_commit() {
...
+
+    let decision_round_wave_2 = 5;
+    let references_decision_round_wave_2 = build_dag(
+        context.clone(),
+        dag_state.clone(),
+        None,
+        decision_round_wave_2,
+    );
+
+    // Commit one block.
```
Comment (mwtian) [inline consensus/core/src/tests/universal_committer_tests.rs:83]:
```suggestion
    // Commit one leader.
```

File: consensus/core/src/tests/universal_committer_tests.rs [hunk 4]
```diff
@@ -57,6 +58,272 @@ fn direct_commit() {
...
+fn idempotence() {
+    telemetry_subscribers::init_for_testing();
+    // Commitee of 4 with even stake
+    let context = Arc::new(Context::new_for_test(4).0);
+    let dag_state = Arc::new(RwLock::new(DagState::new(
+        context.clone(),
+        Arc::new(MemStore::new()),
+    )));
+
+    // Create committer without pipelining and only 1 leader per round
```
Comment (mwtian) [inline consensus/core/src/tests/universal_committer_tests.rs:72]: 1 leader per round or per wave?

Comment (arun-koshy) [reply consensus/core/src/tests/universal_committer_tests.rs]: spoke offline, but 1 leader per leader round to capture that multi leader is disabled.

File: consensus/core/src/tests/universal_committer_tests.rs (hunk 2 above)
Comment (arun-koshy) [reply consensus/core/src/tests/universal_committer_tests.rs:69]: Added the assertions for authority index & round. The leader committed should be the leader of wave 1. The variable names should help describe that.

**Comment by arun-koshy** [comment]
Added additional clarifying comments around the tests to hopefully make it clearer. Though as we discussed offline the setup can still be hard to follow without some additional infrastructure to describe the DAG as a string which can happen in a follow up PR

**Review by mwtian — APPROVED** [review]
